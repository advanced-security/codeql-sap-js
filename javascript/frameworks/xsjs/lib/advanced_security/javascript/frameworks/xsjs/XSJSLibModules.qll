/** Provides classes for working with XSJSLib modules. */

import javascript

/**
 * An XSJS module.
 */
class XSJSModule extends Module {
  XSJSModule() { this.getFile().getExtension() = ["xsjs", "xsjslib"] }

  override XSJSImportExpr getAnImport() { result.getTopLevel() = this }

  /** Gets a module from which this module imports. */
  override XSJSModule getAnImportedModule() { result = this.getAnImport().getImportedModule() }

  override DataFlow::ValueNode getAnExportedValue(string name) {
    result.getFile() = this.getFile() and
    exists(FunctionDeclStmt fds | fds = result.getAstNode() |
      fds.getParent() instanceof TopLevel and
      name = fds.getName()
    )
  }
}

/**
 * An import declaration.
 *
 * Example:
 *
 * ```
 * $.import("module.xsjslib");
 * ```
 */
class XSJSImportExpr extends CallExpr, Import {
  XSJSImportExpr() {
    this.getReceiver().(GlobalVarAccess).getName() = "$" and
    this.getFile().getExtension() = ["xsjs", "xsjslib"] and
    this.getCalleeName() = "import" and
    this.getNumArgument() = 1
  }

  override Module getEnclosingModule() { result = this.getTopLevel() }

  override XSJSLiteralImportPath getImportedPath() { result = this.getArgument(0) }

  override DataFlow::Node getImportedModuleNode() { result = DataFlow::valueNode(this) }
}

/** A literal path expression appearing in an `import` declaration. */
private class XSJSLiteralImportPath extends PathExpr, ConstantString {
  XSJSLiteralImportPath() { exists(XSJSImportExpr req | this = req.getArgument(0)) }

  override Folder getAdditionalSearchRoot(int priority) {
    priority = 0 and
    result = this.getFile().getParentContainer()
  }

  override string getValue() { result = this.getStringValue() }
}
