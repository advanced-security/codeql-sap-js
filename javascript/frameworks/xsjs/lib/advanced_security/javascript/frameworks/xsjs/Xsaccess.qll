import javascript

class ExposedServiceAccessSpec extends File {
  ExposedServiceAccessSpec() {
    this.getBaseName() = "xs-app.json"
    or
    // we are only interested in exposed services
    this.getBaseName() = ".xsaccess" and
    exists(JsonValue v | this = v.getJsonFile() |
      v.getPropValue("exposed").getBooleanValue() = false
    )
  }
}
