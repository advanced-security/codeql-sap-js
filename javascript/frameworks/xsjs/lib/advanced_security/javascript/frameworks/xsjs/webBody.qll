/**
 * Classes to find uses of new $.net.http.Client() .getResponse() calls.
 * A net.http.Client is the object; its .getResponse() method is a data source.
 *
 * @ kind problem
 * @id xsjs-source
 */

import javascript
import DataFlow as DF

// Identify 'new $.net.http.Client'
class NewHTTPClient extends NewExpr {
  NewHTTPClient() {
    exists(DotExpr de1 |
      this.getAChild() = de1 and
      de1.toString() = "$.net.http.Client"
    )
  }
}

// Connect
//     new $.net.http.Client();
//
// to its use:
//     _.getResponse()
DF::SourceNode myType(DF::TypeTracker t) {
  t.start() and
  exists(NewHTTPClient nc | nc.flow().getALocalSource() = result)
  or
  exists(DF::TypeTracker t2 | result = myType(t2).track(t2, t))
}

DF::SourceNode myType() { result = myType(DF::TypeTracker::end()) }

class NetHTTPSource extends MethodCallExpr {
  NetHTTPSource() {
    exists(DF::SourceNode sn, Expr db |
      sn = myType() and
      this.getMethodName() = "getResponse" and
      db = this.getReceiver()
    )
  }
}
