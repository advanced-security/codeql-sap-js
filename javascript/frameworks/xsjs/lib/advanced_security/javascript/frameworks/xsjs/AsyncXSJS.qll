import javascript
import DataFlow
import XSJSLibModules

/**
 * The root XSJS namespace, accessed as a dollar sign (`$`) symbol.
 */
class XSJSDollarNode extends DataFlow::SourceNode {
  XSJSDollarNode() {
    this.accessesGlobal("$") and
    this.getFile().getExtension() = ["xsjs", "xsjslib"]
  }
}

/**
 * `TypeModel` for `XSJSDollarNode`.
 */
class XSJSDollarTypeModel extends ModelInput::TypeModel {
  override DataFlow::Node getASource(string type) {
    type = "XsjsDollar" and
    result = any(XSJSDollarNode dollar)
  }

  /**
   * Prevents model pruning for type `XsjsDollar`
   */
  override predicate isTypeUsed(string type) { type = "XsjsDollar" }
}

/**
 * A reference to either a response or a request. A single-part request or response
 * is accessed through `$.request` or `$.response`, but the multi-part ones are acceesed
 * through `$.request.entities[n]` or `$.response.entities[n]`, respectively.
 */
class XSJSRequestOrResponse extends SourceNode instanceof PropRef {
  string fieldName;

  XSJSRequestOrResponse() {
    fieldName = ["request", "response"] and
    (
      exists(XSJSDollarNode dollar | this = dollar.getAPropertyReference(fieldName))
      or
      exists(XSJSDollarNode dollar |
        this =
          dollar
              .getAPropertyReference(fieldName)
              .(SourceNode)
              .getAPropertyReference("entities")
              .(SourceNode)
              .getAPropertyReference() and
        this.asExpr() instanceof IndexExpr
      )
    )
  }

  predicate isResponse() { fieldName = "response" }

  predicate isRequest() { fieldName = "request" }

  PropRef getHeaders() { result = this.getAPropertyReference("headers") }

  /**
   * Get the immediate node corresponding to the `Content-Type` of this response or request.
   */
  DataFlow::Node getContentType() {
    /* 1. Setting Content-Type via `$.response.contentType` */
    result = this.getAPropertyWrite("contentType").getRhs()
    or
    /* 2. Setting Content-Type via `$.response.headers.set("Content-Type", ...)` */
    exists(MethodCallNode headersSetCall |
      headersSetCall.getReceiver().getALocalSource() = this.getHeaders() and
      headersSetCall.getMethodName() = "set" and
      headersSetCall.getArgument(0).getALocalSource().asExpr().getStringValue() = "Content-Type" and
      result = headersSetCall.getArgument(1)
    )
  }

  /**
   * Holds if the `Content-Type` is scriptable, i.e. able to include a JavaScript source.
   */
  predicate isScriptableContentType() {
    exists(string scriptableMimeType |
      scriptableMimeType in ["text/html", "text/xml", "image/svg+xml"]
    |
      this.getContentType().getALocalSource().asExpr().getStringValue() = scriptableMimeType or
      this.getContentType().asExpr().getStringValue() = scriptableMimeType
    )
  }

  /**
   * Holds if the `Content-Type` is dependent on a remote data.
   */
  predicate contentTypeIsDependentOnRemote() {
    this.getContentType().getALocalSource() instanceof RemoteFlowSource
  }
}

/**
 * A reference to a request. A single-part request is accessed through `$.request`,
 * but a multi-part one is accessed through `$.request.entities[n]`.
 */
class XSJSRequest extends XSJSRequestOrResponse {
  XSJSRequest() { this.isRequest() }

  /**
   * Gets an intraprocedural predecessor or a successor of this request, control-flow wise.
   */
  XSJSRequest getAPredOrSuccRequest() {
    exists(ControlFlowNode cfgNode |
      (
        cfgNode = this.asExpr().getFirstControlFlowNode().getAPredecessor+() or
        cfgNode = this.asExpr().getFirstControlFlowNode().getASuccessor+()
      ) and
      result.asExpr().getFirstControlFlowNode() = cfgNode
    )
  }
}

/**
 * A reference to a response. A single-part request is accessed through `$.response`,
 * but a multi-part one is accessed through `$.response.entities[n]`.
 */
class XSJSResponse extends XSJSRequestOrResponse {
  XSJSResponse() { this.isResponse() }

  /**
   * Gets an intraprocedural predecessor or a successor of this response, control-flow wise.
   */
  XSJSResponse getAPredOrSuccResponse() {
    exists(ControlFlowNode cfgNode |
      (
        cfgNode = this.asExpr().getFirstControlFlowNode().getAPredecessor+() or
        cfgNode = this.asExpr().getFirstControlFlowNode().getASuccessor+()
      ) and
      result.asExpr().getFirstControlFlowNode() = cfgNode
    )
  }
}

/**
 * A reference to a header of a request or a response.
 */
class XSJSRequestOrResponseHeaders extends SourceNode instanceof PropRef {
  XSJSRequestOrResponseHeaders() {
    exists(XSJSRequestOrResponse requestOrResponse |
      this = requestOrResponse.getAPropertyReference("headers")
    )
  }

  /**
   * Given a header property name, gets a call to `$.web.TupelList.set(name, value, options?)`
   * on this reference, which sets a header property to a value.
   */
  MethodCallNode getHeaderSetCall(string name) {
    result.getReceiver().getALocalSource() = this and
    result.getMethodName() = "set" and
    result.getArgument(0).getALocalSource().asExpr().getStringValue() = name
  }

  /**
   * Gets a call to `$.web.TupelList.set(name, value, options?)` on this reference, which sets
   * a header property to a value.
   */
  MethodCallNode getAHeaderSetCall() { result = this.getHeaderSetCall(_) }
}

/**
 * A reference to a database connection obtained either via the deprecated `$.db` namespace
 * or the now-preferred `$.hdb` namespace.
 */
class XSJSDatabaseConnectionReference extends MethodCallNode {
  string subNamespace;

  XSJSDatabaseConnectionReference() {
    exists(XSJSDollarNode dollar |
      this.getMethodName() = "getConnection" and
      this.getReceiver().getALocalSource() = dollar.getAPropertyReference(subNamespace)
    )
  }

  predicate isDbSubNamespace() { subNamespace = "db" }

  predicate isHdbSubNamespace() { subNamespace = "hdb" }

  /**
   * Gets the call that prepares the statement depending on the namespace.
   */
  abstract MethodCallNode getStatementPreparingCall();
}

/**
 * A reference to a database connection obtained via the deprecated `$.db` namespace.
 */
class XSJSDBConnectionReference extends XSJSDatabaseConnectionReference {
  XSJSDBConnectionReference() { this.isDbSubNamespace() }

  override MethodCallNode getStatementPreparingCall() {
    result = this.getAMemberCall("prepareStatement")
  }
}

/**
 * A reference to a database connection obtained via the now-preferred `$.hdb` namespace.
 */
class XSJSHDBConnectionReference extends XSJSDatabaseConnectionReference {
  XSJSHDBConnectionReference() { this.isHdbSubNamespace() }

  override MethodCallNode getStatementPreparingCall() {
    result = this.getAMemberCall("executeQuery")
  }
}

class XSJSUtilNamespace extends SourceNode instanceof PropRef {
  XSJSUtilNamespace() {
    exists(XSJSDollarNode dollar | this = dollar.getAPropertyReference("util"))
  }
}

class XSJSZipInstance extends NewNode {
  XSJSZipInstance() {
    exists(XSJSUtilNamespace util | this = util.getAConstructorInvocation("Zip"))
  }
}
