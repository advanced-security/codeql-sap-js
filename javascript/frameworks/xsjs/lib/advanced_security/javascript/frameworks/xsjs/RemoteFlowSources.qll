import javascript
import DataFlow
import advanced_security.javascript.frameworks.xsjs.AsyncXSJS

class XSJSZipClassInstantiation extends NewNode {
  XSJSZipClassInstantiation() {
    exists(XSJSUtilNamespace util |
      this = util.getAConstructorInvocation("Zip") and
      this.getAnArgument().getALocalSource() instanceof RemoteFlowSource
    )
  }
}

private SourceNode xsjsRequestBody(DataFlow::TypeTracker t) {
  t.start() and
  exists(XSJSRequest dollarRequest | result = dollarRequest.getAPropertyRead("body"))
  or
  exists(DataFlow::TypeTracker t2 | result = xsjsRequestBody(t2).track(t2, t))
}

private SourceNode xsjsRequestBody() { result = xsjsRequestBody(DataFlow::TypeTracker::end()) }

class XSJSRequestBody extends SourceNode {
  XSJSRequestBody() {
    exists(SourceNode dollarRequestBody | dollarRequestBody = xsjsRequestBody() |
      this = dollarRequestBody or
      this = dollarRequestBody.getAMethodCall(["asArrayBuffer", "asString", "asWebRequest"])
    )
  }
}
