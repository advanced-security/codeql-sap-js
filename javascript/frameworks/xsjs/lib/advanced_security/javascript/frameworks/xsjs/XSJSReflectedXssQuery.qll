import javascript
import advanced_security.javascript.frameworks.xsjs.AsyncXSJS
import semmle.javascript.security.dataflow.DomBasedXssQuery as DomBasedXss

class XSJSResponseSetBodyCall extends MethodCallNode {
  XSJSResponse response;

  XSJSResponseSetBodyCall() {
    this.getMethodName() = "setBody" and
    this.getReceiver() = response
  }

  XSJSResponse getParentXSJSResponse() { result = response }
}

class Configuration extends TaintTracking::Configuration {
  Configuration() { this = "XSJS Reflected XSS Query" }

  override predicate isSource(DataFlow::Node start) {
    exists(DomBasedXss::Configuration domBasedXssConfiguration |
      domBasedXssConfiguration.isSource(start)
    ) or
    start instanceof RemoteFlowSource
  }

  override predicate isSink(DataFlow::Node end) {
    exists(XSJSResponseSetBodyCall setBody, XSJSResponse thisOrAnotherXSJSResponse |
      thisOrAnotherXSJSResponse = setBody.getParentXSJSResponse() or
      thisOrAnotherXSJSResponse = setBody.getParentXSJSResponse().getAPredOrSuccResponse()
    |
      end = setBody.getArgument(0) and
      (
        thisOrAnotherXSJSResponse.isScriptableContentType() or
        thisOrAnotherXSJSResponse.contentTypeIsDependentOnRemote()
      )
    )
  }
}
