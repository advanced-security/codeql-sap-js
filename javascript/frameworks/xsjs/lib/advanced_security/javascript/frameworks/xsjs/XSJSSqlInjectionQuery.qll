import javascript
import advanced_security.javascript.frameworks.xsjs.AsyncXSJS
import semmle.javascript.security.dataflow.SqlInjectionQuery as SqlInjection

class XSJSDBConnectionPrepareStatementArgument extends DataFlow::ValueNode {
  XSJSDBConnectionPrepareStatementArgument() {
    exists(XSJSDatabaseConnectionReference connection |
      this = connection.getStatementPreparingCall().getArgument(0)
    )
  }

  predicate isConcatenated() { this.getAPredecessor+() instanceof StringOps::ConcatenationNode }
}

module Configuration implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node start) {
    SqlInjection::SqlInjectionConfig::isSource(start)
    or
    start instanceof RemoteFlowSource
  }

  predicate isSink(DataFlow::Node end) {
    end.(XSJSDBConnectionPrepareStatementArgument).isConcatenated()
  }

  predicate isBarrier(DataFlow::Node node) { SqlInjection::SqlInjectionConfig::isBarrier(node) }

  predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
    SqlInjection::SqlInjectionConfig::isAdditionalFlowStep(node1, node2)
  }
}
