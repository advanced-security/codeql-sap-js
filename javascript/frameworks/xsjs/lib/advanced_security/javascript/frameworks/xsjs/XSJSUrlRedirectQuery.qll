import javascript
import advanced_security.javascript.frameworks.xsjs.AsyncXSJS
import semmle.javascript.security.dataflow.ServerSideUrlRedirectQuery as UrlRedirect

class Configuration extends TaintTracking::Configuration {
  Configuration() { this = "XSJS URL Redirect Query" }

  override predicate isSource(DataFlow::Node start) {
    exists(UrlRedirect::Configuration urlRedirectConfiguration |
      urlRedirectConfiguration.isSource(start)
    ) or
    start instanceof RemoteFlowSource
  }

  override predicate isSink(DataFlow::Node end) {
    exists(XSJSRequestOrResponseHeaders headers |
      end = headers.getHeaderSetCall("location").getArgument(1)
    )
  }
}
