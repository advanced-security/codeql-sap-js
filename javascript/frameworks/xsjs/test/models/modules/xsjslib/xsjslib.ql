import javascript
import advanced_security.javascript.frameworks.xsjs.XSJSLibModules

select any(Module m)