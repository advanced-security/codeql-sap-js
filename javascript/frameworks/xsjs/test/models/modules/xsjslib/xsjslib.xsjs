var injection1=$.import("xsjslib.xsjslib");

injection1.test1(requestParameters);
