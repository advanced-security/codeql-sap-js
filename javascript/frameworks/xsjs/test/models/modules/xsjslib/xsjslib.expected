| xsjslib.xsjs:1:1:4:0 | <toplevel> |
| xsjslib.xsjslib:1:1:15:2 | <toplevel> |
